### PR TITLE
Provenance retrieval route

### PR DESCRIPTION
**Pull request by woodruffw**
~~WIP, not ready for review yet.~~ This has ~~two~~ three functional changes:

* Removes `Provenance.provenance_digest`, since the latest PEP 740 language doesn't include it
* Adds an internal route (`/_/provenance/{filename}/`) for retrieving a file's provenance, if it exists
* Enables provenance persistence 

That route will be accessible, but isn't linked to any of the index representations in this PR. I'll do that with the final PR, which will be built off of #16624.

Signed-off-by: William Woodruff <william@trailofbits.com>

### CODE REVIEW

File: warehouse/config.py [hunk 1]
```diff
@@ -519,12 +519,14 @@ def configure(settings=None):
...
-        default="100 per hour",
-    ),
+    (
+        maybe_set(
+            settings,
+            "warehouse.account.accounts_search_ratelimit_string",
+            "ACCOUNTS_SEARCH_RATELIMIT_STRING",
+            default="100 per hour",
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/config.py:529]: Yeah, looks like it. Fixing!

File: warehouse/routes.py [hunk 2]
```diff
@@ -576,6 +576,15 @@ def includeme(config):
         domain=warehouse,
     )
 
+    # PEP 740 URLs
+    config.add_route(
+        "attestations.provenance",
+        "/_/provenance/{filename}/",
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py:582]: True...perhaps just `/provenance/{filename}`? Or even something opaque like `/provenance/{file-or-provenance-uuid}`.

(My original thinking behind `/_/` is that the PEP specifies that resolution must always go through the referenced URL, i.e., there's guarantee that it remains stable for a particular file. So we could start out on `/_/` and then move to a more permanent location, but maybe that's more effort than it's worth.)

[the remark this answers is not in the document]
Comment (di) [reply warehouse/routes.py:582]: @ewdurbin @miketheman Any thoughts about where this route should live?

[the remark this answers is not in the document]
Comment (miketheman) [reply warehouse/routes.py:582]: 🤔 I'm not 100% up to speed on attestations vs provenance, but why do we need an internal URL?

If requests are meant to be specific to a file, wouldn't we want to traverse to the filename in question? Something like `/project/pytest/8.3.3/files/pytest-8.3.3.tar.gz/provenance/` ?

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py:582]: > 🤔 I'm not 100% up to speed on attestations vs provenance

TL;DR: "provenance" for a file is the set of one or more "attestations" for that file. It's just the JSON structure we serve to end users for their consumption/verification.

> but why do we need an internal URL?

We don't, it can be any URL (with the constraint that it needs to be something we can currently route on the Warehouse side, since provenance lives in the DB instead of object storage for now) 🙂 

> If requests are meant to be specific to a file, wouldn't we want to traverse to the filename in question? Something like `/project/pytest/8.3.3/files/pytest-8.3.3.tar.gz/provenance/` ?

I think something like this would work! I'm not super familiar with how multiple traversals work, though, so I'll need to play around with it.

[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/routes.py:582]: I'm curious if we anticipate any additional resources accessible in a similar way in the future. That might help establish better basis.

I don't approve of hosting them under /project/ as suggested above, since that's currently all "UI" only.

Maybe something like `/metadata/{project_name}/{release}/{filename}/provenance` works? If we ever wanted to host other metadata about a project, release, or file it seems like that would be extensible.

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py]: Okay, I've updated this to use `/metadata/{project_name}/{release}/{filename}/provenance`, PTAL!

File: warehouse/packaging/models.py [hunk 3]
```diff
@@ -694,6 +694,18 @@ def __table_args__(cls):  # noqa
...
+        session: orm.Session = orm.object_session(self)  # type: ignore[assignment]
+
+        try:
+            return (
+                session.query(File)
+                .filter(File.release == self, File.filename == filename)
+                .one()
+            )
+        except NoResultFound:
+            raise KeyError from None
```
Comment (woodruffw) [inline warehouse/packaging/models.py:707]: NB: Needed to add this to allow traversal through `{release}` in routing. It replaces the `FileFactory` that was previously in this PR.

File: warehouse/routes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (di) [reply warehouse/routes.py]: A few thoughts:

* I think "metadata" is overloaded here. This isn't "metadata" in the normal sense of packaging metadata. Also, we might want to expose a new API for that metadata someday, and the presence of this route would be confusing for that (would it be at `/metadata/.../metadata`?)
* I'd like us to have the option to mount HTML-based views at this same route someday (similar to the PEP 691 simple API)
* Also similar to that API, I like that we have different top-level routes that map to different PEP-related APIs

Given that, I propose:
* we put this at `/integrity/{project_name}/{release}/{filename}/provenance`
* we add content-type negotiation for `application/vnd.pypi.integrity.v1+json`

Since we don't have an HTML-based view for PEP 740 at this time, this should 404 for now if the content-type is not set or is `text/html`.

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py]: > Given that, I propose:
> 
> * we put this at `/integrity/{project_name}/{release}/{filename}/provenance`
> * we add content-type negotiation for `application/vnd.pypi.integrity.v1+json`
> 
> Since we don't have an HTML-based view for PEP 740 at this time, this should 404 for now if the content-type is not set or is `text/html`.

SGTM; I'll update here and you can let me know what you think!

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py]: Done -- this now routes through `/integrity/...` and performs basic content negotiation (it has pieces in place to negotiate HTML, but currently rejects those/only accepts JSON).

File: warehouse/attestations/views.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    if request.flags.enabled(AdminFlagValue.DISABLE_PEP740):
+        return HTTPForbidden(json={"message": "Attestations temporarily disabled"})
+
+    if not file.provenance:
+        return HTTPNotFound(
+            json={"message": f"No provenance available for {file.filename}"}
+        )
+
+    # Apply CORS headers.
+    request.response.headers.update(_CORS_HEADERS)
```
Comment (woodruffw) [inline warehouse/attestations/views.py:71]: NB: This applies the same CORS headers as the `simple_detail` views do, to allow use from within the browser.

File: warehouse/attestations/views.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+@view_config(
+    route_name="attestations.provenance",
+    context=File,
+    require_methods=["GET"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+    decorator=[
+        add_vary("Accept"),
```
Comment (woodruffw) [inline warehouse/attestations/views.py:52]: NB: I didn't add any caching decorators here, yet.

Comment (di) [reply warehouse/attestations/views.py:52]: I _think_ we can go ahead and give this the same cache configuration as `/simple`.

**Review by di — APPROVED** [review]

File: warehouse/api/integrity.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+from warehouse.packaging.models import File
+from warehouse.utils.cors import _CORS_HEADERS
+
+MIME_TEXT_HTML = "text/html"
+MIME_PYPI_INTEGRITY_V1_HTML = "application/vnd.pypi.integrity.v1+html"
+MIME_PYPI_INTEGRITY_V1_JSON = "application/vnd.pypi.integrity.v1+json"
+
+
+def _select_content_type(request: Request) -> str:
+    offers = request.accept.acceptable_offers(
```
Comment (miketheman) [inline warehouse/api/integrity.py:29]: thought: unrelated to this change, I wonder if this pattern is relatively easy to replace with our use of deprecated `.best_match`

File: warehouse/api/integrity.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    if not file.provenance:
+        return HTTPNotFound(
+            json={"message": f"No provenance available for {file.filename}"}
+        )
+
+    # Apply CORS headers.
+    request.response.headers.update(_CORS_HEADERS)
+
+    return file.provenance.provenance
```
Comment (miketheman) [inline warehouse/api/integrity.py:80]: Seeing this made me double-take, since I thought it must be a typo. Is there a better name for `Provenance.provenance`, maybe `Provenance.attestations_json` since that's what they contain? 

Then this line would read:

```python
return file.provenance.attestations_json
```

Is the naming confusing, or is this the normal naming patterns of this universe?

**Review by miketheman — COMMENTED** [review]
I wasn't super familiar with this, so it took me a bit longer than I expected, but maybe that's a good thing!

I left in a bunch of notes, let me know if anything doesn't make sense.

File: tests/functional/forklift/test_legacy.py [hunk 8]
```diff
@@ -270,3 +278,78 @@ def test_invalid_classifier_upload_error(webtest):
...
+
+
+def test_provenance_upload(webtest):
+    user = UserFactory.create(
+        password=(  # 'password'
+            "$argon2id$v=19$m=1024,t=6,p=6$EiLE2Nsbo9S6N+acs/beGw$ccyZDCZstr1/+Y/1s3BVZ"
+            "HOJaqfBroT0JCieHug281c"
+        )
+    )
+    EmailFactory.create(user=user, verified=True)
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply tests/functional/forklift/test_legacy.py:290]: TIL, thanks!

File: warehouse/api/integrity.py (hunk 7 above)
Comment (woodruffw) [reply warehouse/api/integrity.py:80]: Yeah, this naming is non-ideal. The reason I named it `provenance` instead of `attestations_json` is because a "provenance object" is more than just a set of "attestation objects" in PEP 740 (it also has groupings/associated identity information).

WDYT about `provenance_json`? If that seems reasonable to you, I'll pin this for a follow-up PR (since it'll require a small migration).

File: warehouse/config.py [hunk 9]
```diff
@@ -525,7 +525,7 @@ def configure(settings=None):
         "warehouse.account.accounts_search_ratelimit_string",
         "ACCOUNTS_SEARCH_RATELIMIT_STRING",
         default="100 per hour",
-    ),
+    )
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/config.py:528]: I _think_ not -- I tried this in the REPL, and it looks like it results in the return being a tuple, but the config itself is produced as a side effect of `maybe_set` rather than the return value.

For example:

```python
>>> def foo():
...     ...
... 
>>> foo(
...     
... ),
(None,)
```

(TIL this is legal at all -- I know about trailing commas for tuple creation during assignment, but I didn't know it could also be done after a bare function call!)

File: warehouse/routes.py [hunk 10]
```diff
@@ -576,6 +576,15 @@ def includeme(config):
         domain=warehouse,
     )
 
+    # PEP 740 URLs
+    config.add_route(
+        "attestations.provenance",
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/routes.py]: 6bab11141..9fea45791

File: warehouse/api/integrity.py (hunk 7 above)
Comment (miketheman) [reply warehouse/api/integrity.py:80]: Well, a column rename from `provenance` to `provenance_json` doesn't seem to be worth the churn, so it's probably fine as is.

Although if and when you might need to do any migrations related to the Provenance model, please add `comment="..."` to the `mapped_columm` definition, so at the very least the comments appear in the database as well. 

See examples for a table itself as well as all the columns here:
https://github.com/pypi/warehouse/blob/7273c870bafe5d677d57dfdcbe70c86e5ab53a79/warehouse/ip_addresses/models.py#L31-L67

File: warehouse/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (miketheman) [reply warehouse/config.py:528]: Right right - `maybe_set` is mutating the passed in `settings`, and does nothing with the return value.
So not an issue here, but definitely something to keep in mind - since a stray comma can bork an expected response:

```python
>>> def foo():
...    ...
...
>>> r1 = foo()
>>> r2 = foo(),
>>> r1 == r2
False
```

**Review by miketheman — APPROVED** [review]
